File: 2022/Day4/solution.py
def part1(_input: str) -> int:
	with open(_input, "r", encoding="UTF-8") as file:
		return sum(rangeIsSubset(*map(int, line.replace(",", "-").split("-"))) for line in file.readlines())

def part2(_input: str) -> int:
	with open(_input, "r", encoding="UTF-8") as file:
		return sum(rangeOverlap(*map(int, line.replace(",", "-").split("-"))) for line in file.readlines())

def rangeIsSubset(start1: int, end1: int, start2: int, end2: int) -> bool:
	return start1 == start2 or end1 == end2 or (start1 > start2) == (end1 < end2)

def rangeOverlap(start1: int, end1: int, start2: int, end2: int) -> bool:
	return start1 <= end2 and start2 <= end1

_input = "./2022/Day4/input.txt"

print("Part 1:", part1(_input))
print("Part 2:", part2(_input))

File: 2022/Day4/solution.c
#include <stdio.h>
#include <stdbool.h>

bool rangeIsSubset(int min1, int max1, int min2, int max2) {
	return 1 > ((min1 - min2) * (max1 - max2));
}

bool rangeOverlap(int min1, int max1, int min2, int max2) {
	return min1 <= max2 && min2 <= max1;
}

int part1() {
	int total = 0;
	int min1, max1, min2, max2;
	while (scanf("%d-%d,%d-%d\n", &min1, &max1, &min2, &max2) != EOF) {
		total += rangeIsSubset(min1, max1, min2, max2);
	}
	printf("Part 1: %d\n", total);
	return 0;
}

int part2() {
	int total = 0;
	int min1, max1, min2, max2;
	while (scanf("%d-%d,%d-%d\n", &min1, &max1, &min2, &max2) != EOF) {
		total += rangeOverlap(min1, max1, min2, max2);
	}
	printf("Part 2: %d\n", total);
	return 0;
}

int main() {
	//part1();
	part2();
	return 0;
}
